// @zendesk/zcli-apps/dist/utils/getAppSettings.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const getAppSettings_1 = require("./getAppSettings");
const createApp = require("./createApp");
const test_1 = require("@oclif/test");
const manifestOutput = {
    name: 'app 1',
    author: {
        name: 'name',
        email: 'test@email.com'
    },
    defaultLocale: 'en',
    location: {
        sell: {
            top_bar: 'assets/iframe.html'
        },
        support: {
            ticket_editor: 'assets/iframe.html',
            nav_bar: 'assets/iframe.html'
        }
    },
    singleInstall: true,
    signedUrls: false,
    frameworkVersion: '2.0',
    parameters: [{
            name: 'someToken',
            type: 'text',
            secure: true
        }, {
            name: 'salesForceId',
            type: 'number',
            secure: false
        }]
};
describe('getAppSettings', () => {
    test_1.test
        .stub(createApp, 'promptAndGetSettings', () => ({ someToken: 'ABC123' }))
        .it('should return setting from config and prompt for missing config', async () => {
        const settings = await (0, getAppSettings_1.getAppSettings)(manifestOutput, { salesForceId: 222 });
        (0, test_1.expect)(settings).to.deep.equals({
            salesForceId: 222,
            someToken: 'ABC123'
        });
    });
    test_1.test
        .it('should return all settings from config', async () => {
        const settings = await (0, getAppSettings_1.getAppSettings)(manifestOutput, { salesForceId: 222, someToken: 'XYZ786' });
        (0, test_1.expect)(settings).to.deep.equals({
            salesForceId: 222,
            someToken: 'XYZ786'
        });
    });
});
